dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs

```diff
@@ -98,95 +99,286 @@ public Plan CreatePlan(string goal)
     public async Task<SKContext> ExecutePlanAsync(
         [Description("The question to answer")]
         string question,
-        [Description("List of tool descriptions")]
-        string functionDescriptions,
-        SKContext context)
+        SKContext context,
+        CancellationToken token = default)
     {
+        if (string.IsNullOrEmpty(question))
+        {
+            context.Variables.Update("Question not found.");
+            return context;
+        }
+
         var stepsTaken = new List<SystemStep>();
-        if (!string.IsNullOrEmpty(question))
+
+        ChatHistory chatHistory = await this.InitializeChatHistoryAsync(this.CreateChatHistory(this._kernel, out var aiService), aiService, context).ConfigureAwait(false);
+        var chatCompletion = aiService as IChatCompletion;
+        var textCompletion = aiService as ITextCompletion;
+
+        var startingMessageCount = chatHistory.Messages.Count;
+
+        var GetMessageTokens = () =>
+        {
+            var messages = string.Join("\n", chatHistory?.Messages);
+            var tokenCount = messages.Length / 4;
+            return tokenCount;
+        };
+
+        var GetCompletionAsync = async () =>
         {
-            for (int i = 0; i < this.Config.MaxIterations; i++)
+            if (chatCompletion is not null)
+            {
+                var llmResponse = (await chatCompletion.GenerateMessageAsync(chatHistory, ChatRequestSettings.FromCompletionConfig(this._promptConfig.Completion), token).ConfigureAwait(false));
+                return llmResponse;
+            }
+            else if (textCompletion is not null)
             {
-                var scratchPad = this.CreateScratchPad(question, stepsTaken);
+                var thoughtProcess = string.Join("\n", chatHistory.Messages.Select(m => m.Content));
+                var results = (await textCompletion.GetCompletionsAsync(thoughtProcess, CompleteRequestSettings.FromCompletionConfig(this._promptConfig.Completion), token).ConfigureAwait(false));
+
+                if (results.Count == 0)
+                {
+                    throw new SKException("No completions returned.");
+                }
 
-                context.Variables.Set("agentScratchPad", scratchPad);
+                return await results[0].GetCompletionAsync(token).ConfigureAwait(false);
+            }
+
+            throw new SKException("No AIService available for getting completions.");
+        };
```

Comment:
Having a wrapper around the text completion as in my PR feels like the cleaner solution to me. This works, but risks having the code accidentally diverge in the future.

If you don't want to implement a full fledged wrapper you could implement it as a private class internal to StepwisePlanner for now and consider making it public in the future.

Reply:
cc @RogerBarreto if they have any thoughts on creating that kind of wrapper, #2444  for context.

I'm not sure if we want to be prescriptive or not when passing a chathistory to a text completion endpoint. And while we could leverage such an interface to allow users to use StepwisePlanner with their own converters, I don't know if that's the stance we want to take with the kernel. These planners are extensions of the kernel and demonstrations of what can be built using kernel completion services and skill collections. And while we aim for them to be useful for many scenarios, we admit we are not going to be able to build a single planner that is capable of every scenario.

Reply:
Definitely open to incremental improvements here, too, so opening #2640 to track this.

Reply:
Makes sense to decouple. I'll take a stab at that wrapper because it turns out I may need it soon.

If callers want their own implementation they can always write their own wrapper and register it as a chat completion and it will be picked up by this planner.